### PR TITLE
Electrum integration for Bitcoin chain

### PR DESCRIPTION
**Pull request by nkuba**
This PR implements Bitcoin package integration with [Electrum API](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol.html).

There are a couple of servers implementations supporting the Electrum API:
- [ElectrumX](https://github.com/spesmilo/electrumx)
- [Esplora - Electrs by Blockstream](https://github.com/Blockstream/electrs)
- [Fulcrum](https://github.com/cculianu/Fulcrum)

A list of publicly exposed Electrum protocol servers for Bitcoin can be found here:
- mainnet: https://1209k.com/bitcoin-eye/ele.php?chain=btc
- testnet: https://1209k.com/bitcoin-eye/ele.php?chain=tbtc

We added integration tests to verify the implementation works correctly with the most popular server implementations. To run the integration tests execute:
```
go test -v -tags=integration ./...
```

There are a couple of TODOs in the code regarding the improvements that will be handled in follow-up PRs.

Refs https://github.com/keep-network/keep-core/pull/3384
Refs https://github.com/keep-network/keep-core/issues/3369
Refs https://github.com/keep-network/keep-core/pull/3395

### CODE REVIEW

File: pkg/internal/byteutils/byteutils.go [hunk 1]
```diff
@@ -22,3 +22,12 @@ func LeftPadTo32Bytes(bytes []byte) ([]byte, error) {
 
 	return result, nil
 }
+
+// Reverse reverses bytes order in the slice.
```
Comment (tomaszslabon) [inline pkg/internal/byteutils/byteutils.go:26]: I would add information to the description that the function returns a new byte array without modifying the input array.

File: pkg/bitcoin/electrum/electrum_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+	"time"
+
+	"github.com/go-test/deep"
+	"golang.org/x/exp/slices"
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+)
+
+// TODO: Include integration test in the CI.
+// To run the tests execute `go test -v -tags=integration ./...`
```
Comment (tomaszslabon) [inline pkg/bitcoin/electrum/electrum_integration_test.go:21]: When I tried to run this command, I got:
```
pkg/bitcoin/electrum/electrum_integration_test.go:14:2: no required module provides package github.com/go-test/deep; to add it:
        go get github.com/go-test/deep
```

After running  `go get github.com/go-test/deep`, the package got installed and `go.mod` and `go.sum` got updated.

We should probably add this package to `go.mod` and `go.sum` in this PR as well.

Comment (tomaszslabon) [inline pkg/bitcoin/electrum/electrum_integration_test.go:21]: Also, I'm getting those kind of errors:
```
--- FAIL: TestGetLatestBlockHeight_Integration (2.34s)
    --- FAIL: TestGetLatestBlockHeight_Integration/electrs-esplora_ssl (0.31s)
        electrum_integration_test.go:290: failed to initialize electrum client: [EOF]
--- FAIL: TestGetBlockHeader_Integration (3.09s)
    --- FAIL: TestGetBlockHeader_Integration/electrs-esplora_tcp (1.15s)
        electrum_integration_test.go:290: failed to get server version: [retry timeout [1s] exceeded; most recent error: [ServerVersion failed: [request timeout]]]
    --- FAIL: TestGetBlockHeader_Integration/electrs-esplora_ssl (0.31s)
        electrum_integration_test.go:290: failed to initialize electrum client: [EOF]
--- FAIL: TestGetBlockHeader_Negative_Integration (4.65s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrs-esplora_tcp (1.16s)
        electrum_integration_test.go:290: failed to get server version: [retry timeout [1s] exceeded; most recent error: [ServerVersion failed: [request timeout]]]
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrs-esplora_ssl (0.31s)
        electrum_integration_test.go:290: failed to initialize electrum client: [read tcp 192.168.2.102:60812->35.225.54.191:60002: read: connection reset by peer]
```

Comment (nkuba) [reply pkg/bitcoin/electrum/electrum_integration_test.go:21]: 79a8cdb39aa96a7911ea05c0f9380a1f98243df0

Comment (nkuba) [reply pkg/bitcoin/electrum/electrum_integration_test.go:21]: Unfortunately, the connection to servers is not very stable. 
For the requests, we've got retries, which work pretty well. But we still need to add retries for connection establishment and reconnects on connection drops. I left TODOs in the code:
https://github.com/keep-network/keep-core/blob/79a8cdb39aa96a7911ea05c0f9380a1f98243df0/pkg/bitcoin/electrum/electrum.go#L48
https://github.com/keep-network/keep-core/blob/79a8cdb39aa96a7911ea05c0f9380a1f98243df0/pkg/bitcoin/electrum/electrum.go#L116

We will handle those in a follow-up PR.

Please try re-running the electrum integration tests and see if you will be able to get the failing tests passed at some point.
```
go test -v -tags=integration ./pkg/bitcoin/electrum
```

Comment (tomaszslabon) [reply pkg/bitcoin/electrum/electrum_integration_test.go:21]: I ran the tests, but some of them still fail.
Hope the connection retires will solve the problem.

File: pkg/bitcoin/electrum/electrum_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
```
Comment (tomaszslabon) [inline pkg/bitcoin/electrum/electrum_integration_test.go:21]: @nkuba I ran the integration tests a couple of time, but still getting quite a few errors:
```
tslabon@MacBook-Pro-Tomasz ~/projekt/keep-core $% go test -v -tags=integration ./pkg/bitcoin/electrum
=== RUN   TestGetTransaction_Integration
=== RUN   TestGetTransaction_Integration/electrs-esplora_tcp
=== RUN   TestGetTransaction_Integration/electrs-esplora_ssl
=== RUN   TestGetTransaction_Integration/electrumx_ssl
=== RUN   TestGetTransaction_Integration/fulcrum_ssl
--- PASS: TestGetTransaction_Integration (2.91s)
    --- PASS: TestGetTransaction_Integration/electrs-esplora_tcp (0.46s)
    --- PASS: TestGetTransaction_Integration/electrs-esplora_ssl (0.64s)
    --- PASS: TestGetTransaction_Integration/electrumx_ssl (0.97s)
    --- PASS: TestGetTransaction_Integration/fulcrum_ssl (0.84s)
=== RUN   TestGetTransaction_Negative_Integration
=== RUN   TestGetTransaction_Negative_Integration/electrs-esplora_tcp
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransaction_Negative_Integration/electrs-esplora_ssl
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransaction_Negative_Integration/electrumx_ssl
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
=== RUN   TestGetTransaction_Negative_Integration/fulcrum_ssl
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
--- FAIL: TestGetTransaction_Negative_Integration (6.08s)
    --- FAIL: TestGetTransaction_Negative_Integration/electrs-esplora_tcp (1.30s)
    --- FAIL: TestGetTransaction_Negative_Integration/electrs-esplora_ssl (1.49s)
    --- FAIL: TestGetTransaction_Negative_Integration/electrumx_ssl (1.67s)
    --- FAIL: TestGetTransaction_Negative_Integration/fulcrum_ssl (1.61s)
=== RUN   TestGetTransactionConfirmations_Integration
=== RUN   TestGetTransactionConfirmations_Integration/electrs-esplora_tcp
=== RUN   TestGetTransactionConfirmations_Integration/electrs-esplora_ssl
=== RUN   TestGetTransactionConfirmations_Integration/electrumx_ssl
=== RUN   TestGetTransactionConfirmations_Integration/fulcrum_ssl
--- PASS: TestGetTransactionConfirmations_Integration (4.23s)
    --- PASS: TestGetTransactionConfirmations_Integration/electrs-esplora_tcp (0.81s)
    --- PASS: TestGetTransactionConfirmations_Integration/electrs-esplora_ssl (0.95s)
    --- PASS: TestGetTransactionConfirmations_Integration/electrumx_ssl (1.25s)
    --- PASS: TestGetTransactionConfirmations_Integration/fulcrum_ssl (1.21s)
=== RUN   TestGetTransactionConfirmations_Negative_Integration
=== RUN   TestGetTransactionConfirmations_Negative_Integration/fulcrum_ssl
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
=== RUN   TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_tcp
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_ssl
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransactionConfirmations_Negative_Integration/electrumx_ssl
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
--- FAIL: TestGetTransactionConfirmations_Negative_Integration (6.18s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/fulcrum_ssl (1.58s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_tcp (1.31s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_ssl (1.48s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/electrumx_ssl (1.81s)
=== RUN   TestGetLatestBlockHeight_Integration
=== RUN   TestGetLatestBlockHeight_Integration/fulcrum_ssl
=== RUN   TestGetLatestBlockHeight_Integration/electrs-esplora_tcp
    electrum_integration_test.go:204: failed to subscribe for headers: [retry timeout [1s] exceeded; most recent error: [request failed: [failed to get the blocks tip height: [request timeout]]]]
=== RUN   TestGetLatestBlockHeight_Integration/electrs-esplora_ssl
=== RUN   TestGetLatestBlockHeight_Integration/electrumx_ssl
--- FAIL: TestGetLatestBlockHeight_Integration (46.47s)
    --- PASS: TestGetLatestBlockHeight_Integration/fulcrum_ssl (0.76s)
    --- FAIL: TestGetLatestBlockHeight_Integration/electrs-esplora_tcp (1.31s)
    --- PASS: TestGetLatestBlockHeight_Integration/electrs-esplora_ssl (43.42s)
    --- PASS: TestGetLatestBlockHeight_Integration/electrumx_ssl (0.98s)
=== RUN   TestGetBlockHeader_Integration
=== RUN   TestGetBlockHeader_Integration/electrs-esplora_tcp
=== RUN   TestGetBlockHeader_Integration/electrs-esplora_ssl
=== RUN   TestGetBlockHeader_Integration/electrumx_ssl
=== RUN   TestGetBlockHeader_Integration/fulcrum_ssl
--- PASS: TestGetBlockHeader_Integration (2.79s)
    --- PASS: TestGetBlockHeader_Integration/electrs-esplora_tcp (0.48s)
    --- PASS: TestGetBlockHeader_Integration/electrs-esplora_ssl (0.67s)
    --- PASS: TestGetBlockHeader_Integration/electrumx_ssl (0.82s)
    --- PASS: TestGetBlockHeader_Integration/fulcrum_ssl (0.82s)
=== RUN   TestGetBlockHeader_Negative_Integration
=== RUN   TestGetBlockHeader_Negative_Integration/electrs-esplora_tcp
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [missing header]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [missing header]]]
=== RUN   TestGetBlockHeader_Negative_Integration/electrs-esplora_ssl
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [missing header]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [missing header]]]
=== RUN   TestGetBlockHeader_Negative_Integration/electrumx_ssl
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
=== RUN   TestGetBlockHeader_Negative_Integration/fulcrum_ssl
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
--- FAIL: TestGetBlockHeader_Negative_Integration (6.08s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrs-esplora_tcp (1.33s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrs-esplora_ssl (1.51s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrumx_ssl (1.64s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/fulcrum_ssl (1.60s)
=== RUN   TestParseProtocol
=== RUN   TestParseProtocol/lower_case:_tcp
=== RUN   TestParseProtocol/upper_case:_TCP
=== RUN   TestParseProtocol/mixed_case:_Tcp
=== RUN   TestParseProtocol/lower_case:_ssl
=== RUN   TestParseProtocol/upper_case:_SSL
=== RUN   TestParseProtocol/mixed_case:_SsL
=== RUN   TestParseProtocol/invalid:_tls
=== RUN   TestParseProtocol/empty
--- PASS: TestParseProtocol (0.00s)
    --- PASS: TestParseProtocol/lower_case:_tcp (0.00s)
    --- PASS: TestParseProtocol/upper_case:_TCP (0.00s)
    --- PASS: TestParseProtocol/mixed_case:_Tcp (0.00s)
    --- PASS: TestParseProtocol/lower_case:_ssl (0.00s)
    --- PASS: TestParseProtocol/upper_case:_SSL (0.00s)
    --- PASS: TestParseProtocol/mixed_case:_SsL (0.00s)
    --- PASS: TestParseProtocol/invalid:_tls (0.00s)
    --- PASS: TestParseProtocol/empty (0.00s)
FAIL
FAIL    github.com/keep-network/keep-core/pkg/bitcoin/electrum  75.073s
FAIL
```

File: cmd/tbtc_maintainer.go [hunk 4]
```diff
@@ -42,18 +44,18 @@ func tbtcMaintainer(cmd *cobra.Command, args []string) error {
 	ctx := context.Background()
 
 	// TODO: Add connection to the Bitcoin electrum chain:
```
[the remark this answers is not in the document]
Comment (nkuba) [reply cmd/tbtc_maintainer.go]: 776d2aef83772eb3efd827f6b72d1ac391a291fd

File: pkg/bitcoin/electrum/electrum_integration_test.go (hunk 3 above)
Comment (nkuba) [reply pkg/bitcoin/electrum/electrum_integration_test.go:21]: I'm so sorry! I missed pushing the integration test updates. It should work fine now!

File: pkg/bitcoin/electrum/config.go [hunk 5]
```diff
@@ -1,13 +1,41 @@
 package electrum
 
-// BitcoinConfig defines the configuration for Bitcoin.
-type BitcoinConfig struct {
```
[the remark this answers is not in the document]
Comment (nkuba) [reply pkg/bitcoin/electrum/config.go:4]: We discussed it with @tomaszslabon, @pdyraga, and @lukasz-zimnoch and came to the conclusion that moving the `BitcoinConfig` to `pkg/config` is a good solution.

File: pkg/bitcoin/electrum/electrum_integration_test.go (hunk 3 above)
Comment (tomaszslabon) [reply pkg/bitcoin/electrum/electrum_integration_test.go:21]: Ok, It's working now.

**Review by tomaszslabon — APPROVED** [review]

**Pull request merged**